securedrop/journalist_app/main.py

```diff
@@ -74,13 +87,11 @@ def index():
                 starred.append(source)
             else:
                 unstarred.append(source)
-            source.num_unread = len(
-                Submission.query.filter_by(source_id=source.id,
-                                           downloaded=False).all())
+            submissions = Submission.query.filter_by(source_id=source.id).all()
+            unseen_submissions = [s for s in submissions if not s.seen]
+            source.num_unread = len(unseen_submissions)
```

Comment:
I'm sure this calculation could be done in the query, but only if you're curious. I don't think it would save much time.

Reply:
you're right, i can be done as part of the query. i switched to doing it this way after getting an `sqlalchemy.exc.StatementError` when trying to do it this way:

```py
            unseen_submissions = Submission.query.filter_by(source_id=source.id).filter(Submission.seen == false()).all()
```

but now realizing that's because "seen" is not an sql boolean (false() is imported from sqlalchemy.sql.expression) because it's not a column, just a property. so i can just to this instead:

```py
            unseen_submissions = Submission.query.filter_by(source_id=source.id).filter(not Submission.seen).all()
```

Reply:
You're getting tripped up by the leaky abstraction: `not Submission.seen` just evaluates to `false` before getting passed to SQLAlchemy, so this is always returning nothing. The generated query looks like this:

```
SELECT submissions.id, submissions.uuid, submissions.source_id, submissions.filename, submissions.size, submissions.downloaded, submissions.checksum 
FROM submissions 
WHERE false
```

Which returns no unseen submissions, so the sources are all always read.

It would need something like:
```
source.num_unread = (
    Submission.query.filter_by(source_id=source.id)
    .filter(
        and_(
            ~exists().where(SeenFile.file_id == Submission.id),
            ~exists().where(SeenMessage.message_id == Submission.id)
        )
    )
    .count()
)
```

where the query evaluates to:

```
SELECT submissions.id, submissions.uuid, submissions.source_id, submissions.filename, submissions.size, submissions.downloaded, submissions.checksum 
FROM submissions 
WHERE submissions.source_id = :source_id_1 AND NOT (EXISTS (SELECT * 
FROM seen_files 
WHERE seen_files.file_id = submissions.id)) AND NOT (EXISTS (SELECT * 
FROM seen_messages 
WHERE seen_messages.message_id = submissions.id))
```

If you want to push through this, you can use `current_app.logger.error("QUERY: %s", unseen_submissions.statement)` to see what SQLAlchemy is generating.

Reply:
If we want to get too clever by half, we could add methods that return filter expressions: 
```
    @classmethod
    def unseen(cls):
        """
        If a submission has not been downloaded or seen by any journalist, then the submission is considered unseen.
        """
        return and_(
            ~exists().where(SeenFile.file_id == Submission.id),
            ~exists().where(SeenMessage.message_id == Submission.id)
        )
```

used like this:
```
        unseen_submissions += (
            Submission.query.filter_by(source_id=source.id)
            .filter(Submission.unseen())
            .all()
        )
```

Tricksy, but could reduce repetition.

Reply:
since we're code freezing soon, let's go back to the way i had it originally